File: src/main/java/de/jonihoffi/plugins/enhavo/extensions/TemplateCompletionContributor.java
package de.jonihoffi.plugins.enhavo.extensions;

import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.patterns.StandardPatterns;
import com.intellij.util.ProcessingContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.yaml.YAMLLanguage;
import org.jetbrains.yaml.psi.YAMLKeyValue;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import de.jonihoffi.plugins.enhavo.settings.AppSettingsState;

public class TemplateCompletionContributor extends CompletionContributor {
    public TemplateCompletionContributor() {
        extend(
                CompletionType.BASIC,
                PlatformPatterns.psiElement().withLanguage(YAMLLanguage.INSTANCE)
                        .and(PlatformPatterns.psiElement().inside(YAMLKeyValue.class))
                        .and(PlatformPatterns.psiElement().withSuperParent(2, PlatformPatterns.psiElement(YAMLKeyValue.class)
                                .withName(StandardPatterns.string().equalTo("template")))),
                new CompletionProvider<CompletionParameters>() {
                    @Override
                    protected void addCompletions(@NotNull CompletionParameters parameters,
                                                  @NotNull ProcessingContext context,
                                                  @NotNull CompletionResultSet resultSet) {
                        Project project = parameters.getPosition().getProject();
                        String templateDir = project.getBasePath() + "/" + AppSettingsState.getInstance().templatePath;
                        VirtualFile baseDir = LocalFileSystem.getInstance().findFileByPath(templateDir);
                        if (baseDir != null && baseDir.isDirectory()) {
                            for (VirtualFile file : baseDir.getChildren()) {
                                if ("yaml".equals(file.getExtension())) {
                                    LookupElementBuilder lookupElement = LookupElementBuilder.create(file.getNameWithoutExtension());
                                    resultSet.addElement(PrioritizedLookupElement.withPriority(lookupElement, 100.0));
                                }
                            }
                        }
                    }
                }
        );
    }
}

File: src/main/kotlin/de/jonihoffi/plugins/demo/TemplateCompletionContributor.kt
package de.jonihoffi.plugins.demo

import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.patterns.PlatformPatterns
import com.intellij.patterns.StandardPatterns
import com.intellij.util.ProcessingContext
import org.jetbrains.yaml.psi.YAMLKeyValue
import org.jetbrains.yaml.YAMLLanguage

class TemplateCompletionContributor : CompletionContributor() {
    init {
        extend(
                CompletionType.BASIC,
                PlatformPatterns.psiElement().withLanguage(YAMLLanguage.INSTANCE)
                        .and(PlatformPatterns.psiElement().inside(YAMLKeyValue::class.java))
                        .and(PlatformPatterns.psiElement().withSuperParent(2, PlatformPatterns.psiElement(YAMLKeyValue::class.java)
                                .withName(StandardPatterns.string().equalTo("template")))),
                object : CompletionProvider<CompletionParameters>() {
                    override fun addCompletions(
                            parameters: CompletionParameters,
                            context: ProcessingContext,
                            resultSet: CompletionResultSet
                    ) {
                        val project = parameters.position.project
                        val templateDir = project.baseDir.findFileByRelativePath("config/block/templates")
                        if (templateDir != null && templateDir.isDirectory) {
                            templateDir.children.filter { it.extension == "yaml" }
                                    .forEach {
                                        val lookupElement = LookupElementBuilder.create(it.nameWithoutExtension)
                                        val prioritizedElement = PrioritizedLookupElement.withPriority(lookupElement, 100.0)
                                        resultSet.addElement(prioritizedElement)
                                    }
                        }
                    }
                }
        )
    }
}
